astra/src/main/java/com/slack/astra/graphApi/GraphConfig.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    public void setOverrideKey(String overrideKey) {
+      this.overrideKey = overrideKey;
+    }
+  }
+
+  // Holds the entire mapping for logical field names to their configuration of defaults and rules.
+  private Map<String, TagConfig> nodeMetadataTagMapping;
+
+  public Map<String, TagConfig> getNodeMetadataTagMapping() {
+    return nodeMetadataTagMapping;
+  }
+
+  public void setNodeMetadataTagMapping(Map<String, TagConfig> nodeMetadataTagMapping) {
+    this.nodeMetadataTagMapping = nodeMetadataTagMapping;
+  }
+
+  // Loads a GraphConfig from a YAML file on disk. On failure, log and return a null config.
+  public static GraphConfig load(String configFile) throws IOException {
+    if (!configFile.isEmpty()) {
+      try {
+        Path path = Path.of(configFile);
+        String yaml = Files.readString(path);
+
+        ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+        mapper.setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);
+        return mapper.readValue(yaml, GraphConfig.class);
+      } catch (Exception e) {
+        LOG.warn("Failed to read or parse dependency graph config file. Returning null config", e);
+        return null;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Resolves the actual tag value for a given logical field, using the provided span tags. Steps:
+   * 1. Look up the TagConfig for this logical field (e.g. "resource"). 2. Default to using its
+   * defaultKey + defaultValue. 3. If rules are defined: - Iterate through each rule in order. - If
+   * a rule’s field/value condition matches, switch keyToUse to overrideKey. - Continue processing
+   * later rules (last matching rule wins). 4. Finally, look up the chosen key in tags. If missing,
+   * fall back to defaultValue. Note: This logic does not currently support multiple field matches
+   * for a single rule.
+   */
+  public String resolve(Map<String, String> tags, String logicalField) {
+    TagConfig baseCfg = nodeMetadataTagMapping.get(logicalField);
+
+    // If the config doesn't define this logical field, just return from raw tags or fallback.
+    if (baseCfg == null) {
+      return tags.getOrDefault(logicalField, "unknown_" + logicalField);
+    }
+
+    // Start with the default key + value.
+    String keyToUse = baseCfg.getDefaultKey();
+    String defaultValue = baseCfg.getDefaultValue();
+
+    // Apply rules in order. Later matching rules override earlier ones.
+    for (RuleConfig rule : baseCfg.getRules()) {
+      // Get the value of the field defined in the rule.
+      String matchVal = tags.getOrDefault(rule.getField(), "unknown_" + rule.getField());
+      // If the rule value matches, override which key to use for resolution of the logical field
+      // only if it exists in tags, otherwise keep the default.
+      if (rule.getValue().equals(matchVal)) {
+        if (tags.containsKey(rule.getOverrideKey())) {
+          keyToUse = rule.getOverrideKey();
+        }
+      }
+    }
```

Comment:
given the `later matching override`, I'd probably implement this by reversing the list and picking the first one that matches. Using streams, it could look like this.

```suggestion
    // Later rules override earlier ones, so start from the back of the list and use the first one that matches
    // where the rule's override key exists in the document.
    String keyToUse = baseCfg.getRules().reversed().stream()
      .filter(rule ->  rule.getValue().equals(tags.getOrDefault(rule.getField(), "unknown_" + rule.getField()))
      .map(RuleConfig::getOverrideKey)
      .filter(tags::containsKey)
      .findFirst().orElse(baseCfg.getDefaultKey());
```

having a method on Rule for checking the field would make that read cleaner by making the first filter less noisy.

 Also if the resolve happens in a hot path, you might want to do the reversing of the list at config construction time, among other things, but that should probably happen later.